File: Binary-Search/0035-search-insert-position.py
class Solution:
  def searchInsert(self, A: List[int], target: int) -> int:
    lo, hi = 0, len(A) - 1

    while lo <= hi:
      mid = lo + (hi - lo) // 2
      if A[mid] < target:
        lo = mid + 1
      elif A[mid] > target:
        hi = mid - 1
      else:
        return mid

    return lo

File: Binary-Search/0035-search-insert-position.cpp
// Time  : O(logn)
// Space : O(1)
class Solution {
public:
  int searchInsert(vector<int>& A, int x) {
    int lo = 0;
    int hi = A.size();
    while (lo < hi) {
      int mid = (lo + hi)>>1;
      if (A[mid] < x) lo = mid+1;
      else hi = mid;
    }
    return lo;
  }
};